server/api/user/user.model.js

```diff
@@ -90,90 +108,409 @@ function isoDateToTime(isoDate){
-    var today = new Date();
-    today.setHours(0,0,0,0);
+    .virtual('presence')
+    .get(function(){
+        var today = new Date();
+        today.setHours(0,0,0,0);
+        var classYear = global.currentClassYear;
+        if (!'attendanceByYear' in this || this.attendanceByYear == undefined){
+            return 'absent';
+        }
 
-    for (var i = 0;i < this.attendance.length;i++){
-      if (isoDateToTime(this.attendance[i]) == today.getTime()){
-        return "present";
-      }
-    }
-    for (var i = 0;i < this.unverifiedAttendance.length;i++){
-      if (isoDateToTime(this.unverifiedAttendance[i]) == today.getTime()){
-        return "unverified";
-      }
-    }
-    return "absent";
-  })
-  .set(function(status){
-    var today = new Date();
-    today.setHours(0,0,0,0);
-    if (status === "present"){
-      // Make sure user is not unverified for today
-      for (var i = this.unverifiedAttendance.length-1;i >= 0;i--){
-        if (isoDateToTime(this.unverifiedAttendance[i]) == today.getTime()){
-           this.unverifiedAttendance.splice(i,1);
+        for (var i = this.attendanceByYear.length-1; i>=0; i--){
+            if (this.attendanceByYear[i].year.equals(classYear._id)){
+
+                var thisYear = this.attendanceByYear[i];
+                for (var a = thisYear.verified.length-1; a>= 0 ; a--){
+                  if (isoDateToTime(thisYear.verified[a].date) === isoDateToTime(today)){
+                    var thisDate = thisYear.verified[a];
+                    if (!thisDate.bonusDay && !thisDate.smallgroup){
+                        return 'present';
+                    }
+                  }
+                }
+                for (var a = thisYear.unverified.length-1; a>= 0 ; a--){
+                  if (isoDateToTime(thisYear.unverified[a].date) === isoDateToTime(today)){
+                    var thisDate = thisYear.unverified[a];
+                    if (!thisDate.bonusDay && !thisDate.smallgroup){
+                        return 'unverified';
+                    }
+                  }
+                }
+                return 'absent';
+            }
         }
-      }
 
-      // If user already has attendance don't change anything
-      for (var i = 0;i < this.attendance.length;i++){
-        if (isoDateToTime(this.attendance[i]) == today.getTime()){
-          return;
+        return 'absent';
+    })
+    .set(function(status){
+        var today = new Date();
+        if (!'attendanceByYear' in this || this.attendanceByYear == undefined){
+            this.attendanceByYear = [];
         }
-      }
-      this.attendance.push(today);
-    }else if (status === "unverified"){
-      // If user already has attendance remove their attendance
-      for (var i = this.attendance.length-1;i >= 0;i--){
-        if (isoDateToTime(this.attendance[i]) == today.getTime()){
-          this.attendance.splice(i,1);
+        //New style attendence
+        var classYear = global.currentClassYear;
```

Comment:
the addition below these lines is incredibly verbose- i suspect this could have been done in a better way...